Create prefer PathSensitivity.NONE best practice item (#35003)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc

=== Tags
`<<tags_reference.adoc#tag:tasks,#tasks>>`, `<<tags_reference.adoc#tag:inputs-and-outputs,#inputs-and-outputs>>`, `<<tags_reference.adoc#tag:configurations,#configurations>>`

[[default_path_sensitivities]]
== Prefer `@PathSensitivity.NONE` for file inputs and `@PathSensitivity.RELATIVE` for directories

Prefer using link:{javadocPath}/org/gradle/api/tasks/PathSensitivity.html#NONE[`@PathSensitivity.NONE`] when defining file inputs for tasks, and link:{javadocPath}/org/gradle/api/tasks/PathSensitivity.html#RELATIVE[`@PathSensitivity.RELATIVE`] when defining directory inputs.

=== Explanation

Most of the time, tasks only care about the **contents** of their input files, not the files' locations.

> You should annotate `@InputFile` and `@InputFiles` properties as `@PathSensitivity.NONE` unless a task's behavior should change based on the location of the input file(s).

Reviewer: I wouldn't phrase it like this. You can use `@InputFiles` with directory inputs as well. I think a better way to phrase this is to talk about when you have only files as inputs, use `NONE`, if you have directories, use `RELATIVE`.

> You should annotate `@InputFile` and `@InputFiles` properties as `@PathSensitivity.NONE` unless a task's behavior should change based on the location of the input file(s).

Author: Taking into account @lptr and @lkasso 's feedback for this section, how about:

When annotating input properties that contain files (such as `@InputFile` and `@InputFiles` collections) use `@PathSensitivity.NONE`. 

This tells Gradle to consider only a file's content when determining if a task is up-to-date.

For input properties that contain directories (such as `@InputDirectory` and `@InputFiles` collections) use `@PathSensitivity.RELATIVE`.

This tells Gradle to consider the path to a directory relative to the build's root directory, while ignoring its absolute location, for up-to-date checks.

=== Tags
`<<tags_reference.adoc#tag:tasks,#tasks>>`, `<<tags_reference.adoc#tag:inputs-and-outputs,#inputs-and-outputs>>`, `<<tags_reference.adoc#tag:configurations,#configurations>>`

[[default_path_sensitivities]]
== Use `@PathSensitivity.NONE` for file inputs and `@PathSensitivity.RELATIVE` for directories

Use link:{javadocPath}/org/gradle/api/tasks/PathSensitivity.html#NONE[`@PathSensitivity.NONE`] for file inputs and link:{javadocPath}/org/gradle/api/tasks/PathSensitivity.html#RELATIVE[`@PathSensitivity.RELATIVE`] for directory inputs.

=== Explanation

Tasks should generally only care about the **contents** of their input files, not their location on disk.

> When annotating input properties that contain files (such as `@InputFile` and `@InputFiles` collections) use `@PathSensitivity.NONE`.

Reviewer 2: ```suggestion
When annotating file-based input properties (for example, `@InputFile` or `@InputFiles` collections), use `@PathSensitivity.NONE`.
```

> This tells Gradle to consider only the files' contents when determining if a task is up-to-date.

Reviewer 2: ```suggestion
This tells Gradle to ignore the path and only consider the file contents when determining whether a task is up-to-date.
```

> For input properties that contain directories (such as `@InputDirectory` and `@InputFiles` collections) use `@PathSensitivity.RELATIVE`.

Reviewer 2: ```suggestion
For directory-based inputs (for example, `@InputDirectory` or `@InputFiles` collections), use `@PathSensitivity.RELATIVE`.
```

This tells Gradle to also consider only the name of the directory (ignoring its absolute location) and to relativize the paths of all files within that directory to it when doing up-to-date checks.

Using `PathSensitivity.NAME_ONLY` or `@PathSensitivity.ABSOLUTE` is generally incorrect.

> `PathSensitivity.NAME_ONLY` tells Gradle to also consider a file's name in addition to its contents, which is rarely useful.

Reviewer 2: ```suggestion
`PathSensitivity.NAME_ONLY` tells Gradle to consider a file's name in addition to its contents, which is rarely useful.

```

`@PathSensitivity.ABSOLUTE` tells Gradle to consider a file's complete absolute path.
This prevents <<build_cache.adoc#build_cache,Build Cache>> and <<configuration_cache.adoc#config_cache,Configuration Cache>> hits across different machines or directories, making your build non-relocatable.

> This can also cause confusion from different task outcomes when the build is run in different directories.

Reviewer 2: ```suggestion
It can also lead to confusing behavior where the same build produces different task outcomes when run from different directories.
```
